Sources/Testing/Traits/CoverageTrait.swift

```diff
@@ -0,0 +1,241 @@
+    /// across tests.
+    public var isolatesTests: Bool
+
+    /// The directory where coverage files are written.
+    ///
+    /// Defaults to `COVERAGE_OUTPUT_DIR` environment variable or current
+    /// working directory.
+    public var outputDirectory: String
+
+    /// Create a coverage trait.
+    ///
+    /// - Parameters:
+    ///   - isolatesTests: Whether to reset counters before each test.
+    ///   - outputDirectory: Directory for coverage files.
+    public init(
+        isolatesTests: Bool = true,
+        outputDirectory: String? = nil
+    ) {
+        self.isolatesTests = isolatesTests
+        self.outputDirectory = outputDirectory ?? _coverageOutputDirectory
+    }
+
+    public var isRecursive: Bool { true }
+}
+
+// MARK: - TestScoping
+
+extension CoverageTrait: TestScoping {
+    public func scopeProvider(for test: Test, testCase: Test.Case?) -> Self? {
+        // When applied to a test function, provide scope to the test function
+        // itself (not individual test cases) so we can disable parallelization
+        // at the suite/function level.
+        test.isSuite || testCase == nil ? self : nil
+    }
+
+    public func provideScope(
+        for test: Test,
+        testCase: Test.Case?,
+        performing function: @Sendable () async throws -> Void
+    ) async throws {
+        // Skip coverage collection if not enabled
+        guard _coverageEnabled else {
+            try await function()
+            return
+        }
+
+        // Enforce serialization - coverage requires serial execution because
+        // LLVM profile counters are shared global state across all threads.
+        guard var configuration = Configuration.current else {
+            try await function()
+            return
+        }
+        configuration.isParallelizationEnabled = false
+
+        try await Configuration.withCurrent(configuration) {
+            // Reset counters to isolate this test's coverage
+            if isolatesTests {
+                __llvm_profile_reset_counters()
+            }
+
+            // Run the test
+            do {
+                try await function()
+            } catch {
+                // Write coverage even if test fails
+                writeCoverage(for: test, testCase: testCase)
+                throw error
+            }
+
+            // Write coverage for this test
+            writeCoverage(for: test, testCase: testCase)
+        }
+    }
+
+    private func writeCoverage(for test: Test, testCase: Test.Case?) {
+        let filename = coverageFilename(for: test, testCase: testCase)
+
+        filename.withCString { cString in
+            __llvm_profile_set_filename(cString)
+        }
+
+        _ = __llvm_profile_write_file()
+    }
+
+    private func coverageFilename(for test: Test, testCase: Test.Case?) -> String {
+        var name = test.name
+        if let testCase {
+            // Include test case ID for parameterized tests
+            name += "-\(testCase.id)"
```

Comment:
The string form of a test case's ID is not suitable for use as a file name. It may be arbitrarily long (too long to be represented as a filename on some systems) or may contain illegal characters.